src/Roslyn.Diagnostics.Analyzers/Core/AbstractApplyTraitToClass.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Roslyn.Diagnostics.Analyzers
+{
+    public abstract class AbstractApplyTraitToClass<TAttributeSyntax> : CodeRefactoringProvider
+        where TAttributeSyntax : SyntaxNode
+    {
+        protected AbstractApplyTraitToClass()
+        {
+        }
+
+        private protected abstract IRefactoringHelpers RefactoringHelpers { get; }
+
+        protected abstract SyntaxNode? GetTypeDeclarationForNode(SyntaxNode reportedNode);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var attribute = await context.TryGetRelevantNodeAsync<TAttributeSyntax>(RefactoringHelpers).ConfigureAwait(false);
+            if (attribute is null)
+            {
+                // No attribute in context
+                return;
+            }
+
+            var syntaxGenerator = SyntaxGenerator.GetGenerator(context.Document);
+            if (syntaxGenerator.TryGetContainingDeclaration(attribute, DeclarationKind.Method) is null)
+            {
+                // The attribute is not applied to a method
+                return;
+            }
+
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+            var attributeType = semanticModel!.GetTypeInfo(attribute, context.CancellationToken);
+            if (attributeType.Type is not { Name: "TraitAttribute", ContainingNamespace: { Name: "Xunit", ContainingNamespace: { IsGlobalNamespace: true } } })
+                return;
+
+            var location = attribute.GetLocation();
+            context.RegisterRefactoring(
+                CodeAction.Create(
+                    RoslynDiagnosticsAnalyzersResources.ApplyTraitToContainingType,
+                    cancellationToken => ApplyTraitToClassAsync(context.Document, location.SourceSpan, cancellationToken),
+                    nameof(AbstractApplyTraitToClass<TAttributeSyntax>)));
+        }
+
+        private async Task<Document> ApplyTraitToClassAsync(Document document, TextSpan sourceSpan, CancellationToken cancellationToken)
+        {
+            var semanticModel = (await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false))!;
```

Comment:
Semantic model was already retrieved in the caller-side. Consider passing it instead.

Reply:
@mavasani any problem with avoiding the capture and just calling `GetSemanticModelAsync` again?

Reply:
⏱ Will capture state